### ISSUE TITLE
High CPU observed on stressnet with downloader

### ISSUE THREAD

**Issue by JackyWYX** [harmony-one/harmony#3618]
**Describe the bug**

High CPU usage observed on Stressnet node running downloader protocol. Observed CPU usage was about 50%

**Comment by JackyWYX** [comment]
CPU profile did several days ago:

![profile](https://user-images.githubusercontent.com/40162696/112877868-2077d200-907c-11eb-83a8-39234e9e2914.png)

Goroutine profile:

![goroutine](https://user-images.githubusercontent.com/40162696/112877883-2372c280-907c-11eb-8430-528ac48c6527.png)

**Comment by JackyWYX** [comment]
From the pprof diagram, we can come to the following conclusion:

1. From CPU profile diagram, we can see that most CPU used is for goroutine coordination.
2. From goroutine diagram, it seems that the node is opening ~124 connection to remote nodes via sync stream.

The root cause is probably the wrong usage of `stream.Close`. In this case, `stream.Reset` might be more appropriate.

Looking deeper into the implementation of these two functions.

**Comment by JackyWYX** [comment]
After running for a day, still observed ~50% cpu usage on non-leader node 54.214.183.56. The number of goroutine running `Stream.Read` is still high.

![goroutine](https://user-images.githubusercontent.com/40162696/113017633-82940e00-9134-11eb-878e-5255eff34532.png)

For CPU profile, `runtime.SiftdownTimer` has greatly reduced. But still not totally sure why CPU is high.

![profile](https://user-images.githubusercontent.com/40162696/113023563-a0fd0800-913a-11eb-8abc-a76ccd00f65b.png)

**Comment by JackyWYX** [comment]
Comparing with a freshly started node, CPU profile looks almost the same:

![profile](https://user-images.githubusercontent.com/40162696/113024153-46b07700-913b-11eb-9135-2c736cb0adcc.png)

**Comment by JackyWYX** [comment]
And the large number of goroutines is also observed on a freshly started node, meaning that it is most likely come from libp2p basic discovery.

**Comment by JackyWYX** [comment]
To this point, we cannot gather any information useful from pprof. Also noticed that mainnet node also have CPU and memory issues after long run. 

So here a baseline on main branch will be made for comparison.

**Comment by JackyWYX** [comment]
When deployed with none-downloader code, the CPU remains < 10% after two days run.

Following is the pprof graph:

![profile](https://user-images.githubusercontent.com/40162696/113333967-f3276000-92d7-11eb-9aed-b560c219560f.png)

![goroutine](https://user-images.githubusercontent.com/40162696/113333981-f6225080-92d7-11eb-995d-e7c37273e9ba.png)

**Comment by JackyWYX** [comment]
Deployed another binary to stressnet with downloader enabled. This time, I will do an extra heap profile to see whether there are some unnecessary memory allocation that cause the CPU usage high.

Also, unstable consensus is disabled to have a comparison against the dns sync. Let it run one day and let's see.

**Comment by JackyWYX** [comment]
CPU issue not observed after 24 hours. Keeps waiting.

**Comment by JackyWYX** [comment]
Observed CPU rate increase in stn at ~1.5 days running. Nothing observed from pprof.

1. Cgo call of signature verification from consensus takes ~50% CPU usage, which is normal.
2. From heap graph, the program takes about 333MB memory, and mostly spent on ipfsDHT (30%) and leveldb (20%), which is normal.

Will need some help and insights from peers.

**Comment by JackyWYX** [comment]
Just one idea flashed in my head: All CPU intensive happens on shard 1 instead of shard 0. Thus I did a pprof on s0 vs s1 node. Here is the heap result:

Shard 0:
![heap](https://user-images.githubusercontent.com/40162696/113614692-b1fcbc00-9607-11eb-93cb-9d0f02fefd3f.png)

Shard 1:
![heap](https://user-images.githubusercontent.com/40162696/113614725-b7f29d00-9607-11eb-8131-791f29f49640.png)

From the graph, I discovered that s1 heap has `time.NewTicker` taking ~5% memory resource which shall not be the normal case. Taking a look into `time.Tick`, it is obvious that this might cause the CPU leak since each time it is called will create `time.NewTicker` which will run its own goroutine forever and is never garbage collected. Change the usage to `time.NewTicker` shall resolve the issue. 

Let's see whether the CPU issue is fixed after 2~3 days observation.
